pkg/services/folder/folderimpl/folder.go

```diff
@@ -456,12 +456,64 @@ func (s *Service) Move(ctx context.Context, cmd *folder.MoveFolderCommand) (*fol
 		return nil, err
 	}
 
+	// here we get the folder, we need to get the height of current folder
+	// and the depth of the new parent folder, the sum can't bypass 8
+	folderHeight, err := s.getHeight(ctx, foldr.UID, cmd.OrgID, &cmd.NewParentUID)
+	if err != nil {
+		return nil, err
+	}
+	parents, err := s.GetParents(ctx, &folder.GetParentsQuery{UID: cmd.NewParentUID, OrgID: cmd.OrgID})
+	if err != nil {
+		return nil, err
+	}
+
+	// current folder height + current folder + parent folder + parent folder depth should be less than or equal 8
+	if folderHeight+len(parents)+2 > folder.MaxNestedFolderDepth {
+		return nil, folder.ErrMaximumDepthReached
+	}
+
+	// if the current folder is already a parent of newparent, we should return error
+	for _, parent := range parents {
+		if parent.UID == foldr.UID {
+			return nil, folder.ErrCircularReference
+		}
+	}
+
+	// if the new parent is the same as the current parent, we don't need to do anything
+	if foldr.ParentUID == cmd.NewParentUID {
+		return foldr, nil
+	}
+
 	return s.store.Update(ctx, folder.UpdateFolderCommand{
 		Folder: foldr,
-		// NewParentUID: &cmd.NewParentUID,
 	})
 }
 
+// getHeight is used to detect the height of the folder, when parentUID is not empty, it will verify that parentUID doesn't exist in the subtree
+func (s *Service) getHeight(ctx context.Context, foldrUID string, orgID int64, parentUID *string) (int, error) {
+	height := -1
+	queue := []string{foldrUID}
+	for len(queue) > 0 {
+		length := len(queue)
+		height++
+		for i := 0; i < length; i++ {
```

Comment:
I think you can write  `for i := range queue` here

Reply:
for the bfs tree traverse i can't loop on the queue entirely since the length of the queue will change dynamically, i need to traverse only for the breath of the level (the length at the beginning of the outer for loop).